### DISCUSSION TITLE
Comparing collections using `Assert.Equal` in 2.5.0 fails, but works in 2.4.2

### DISCUSSION THREAD

**Discussion by toore** [xunit/xunit#2765] [Question] [answered]
Comparing collections using `Assert.Equal` in 2.5.0 fails, when the compared collections contains tuples with arrays.

Let's give an example:
``` C#
Assert.Equal(
    new[]
        {
            (1, new byte[] { 2 })
        },
    new[]
        {
            (1, new byte[] { 2 })
        });
```

In 2.4.2, the above assertion succeeds. I noticed it after upgrading to 2.5.0 when my unit tests were failing. I can update/change my asserts, but it got me curious. Running .NET 7.0.

Is this an unintentional change of the behavior for `Assert.Equal`, or maybe the result of a more correct behavior?

Thanks in advance for any help or guidance. 

Repro solution attached: [CollectionAssertOfTuplesFails.zip](https://github.com/xunit/xunit/files/12330247/CollectionAssertOfTuplesFails.zip)

**Comment by bradwilson** [comment]
No, pretty sure this was unintentional. I'll take a look and see what's going on.

**Reply by bradwilson** [reply, marked as answer]
So, it turns out, this actually is a bug in 2.4.2 and your test was passing accidentally. 😬 

Consider the following two tests:

```csharp
public class UnitTest
{
    [Fact]
    public void Tuple()
    {
        var expected = (1, new[] { 2 });
        var actual   = (1, new[] { 2 });

        Assert.Equal(expected, actual);
    }

    [Fact]
    public void TupleInCollection()
    {
        var expected = new[] { (1, new[] { 2 }) };
        var actual   = new[] { (1, new[] { 2 }) };

        Assert.Equal(expected, actual);
    }
}
```

Here are the results for 2.4.2:

```
xUnit.net Console Runner v2.4.2+f110e5bee5 (64-bit Desktop .NET 4.7.2, runtime: 4.0.30319.42000)
  Discovering: empty (app domain = on [shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  empty (found 2 test cases)
  Starting:    empty (parallel test collections = off, max threads = 24)
    UnitTest.Tuple [FAIL]
      Assert.Equal() Failure
      Expected: (1, System.Int32[])
      Actual:   (1, System.Int32[])
      Stack Trace:
        UnitTest.cs(11,0): at UnitTest.Tuple()
  Finished:    empty
=== TEST EXECUTION SUMMARY ===
   empty  Total: 2, Errors: 0, Failed: 1, Skipped: 0, Time: 0.216s
```

And the results for 2.5.1-pre.22:

```
xUnit.net Console Runner v2.5.1-pre.22+c6073bb042 (64-bit .NET Framework 4.7.2, runtime: 4.0.30319.42000)
  Discovering: empty (app domain = on [shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  empty (found 2 test cases)
  Starting:    empty (parallel test collections = off, max threads = 24)
    UnitTest.TupleInCollection [FAIL]
      Assert.Equal() Failure: Collections differ
                 ↓ (pos 0)
      Expected: [Tuple (1, [2])]
      Actual:   [Tuple (1, [2])]
                 ↑ (pos 0)
      Stack Trace:
        UnitTest.cs(20,0): at UnitTest.TupleInCollection()
    UnitTest.Tuple [FAIL]
      Assert.Equal() Failure: Values differ
      Expected: Tuple (1, [2])
      Actual:   Tuple (1, [2])
      Stack Trace:
        UnitTest.cs(11,0): at UnitTest.Tuple()
  Finished:    empty
=== TEST EXECUTION SUMMARY ===
   empty  Total: 2, Errors: 0, Failed: 2, Skipped: 0, Time: 0.295s
```

The reason for this is that `ValueTuple` (the backing type for anonymous tuples like in the source code above) implements `IEquatable<T>`. In the case of 2.4.2, when the tuple in question is embedded inside a collection, we take a bad path which incorrectly considers the values to be equal; in the case of 2.5.0, the path for comparing the two tuples is the same, regardless of whether it's embedded inside a collection.

To be clear: the `ValueTuple` class is the one that claims to be the authority for equality, and it says they're not equal. Without diving into the source, I can guess with 99.9% accuracy 😂 that the problem is that it doesn't have any special logic for collections, and just considers the two _references_ to be unequal because they point to two different _instances_ of integer arrays (that coincidentally have the same values inside them). This is a reasonable implementation on the part of `ValueTuple`, because its implementation for equality wasn't designed for unit testing. 😄 

-----

That said, there is a new assertion which you could use: `Assert.Equivalent`. Its purpose is to provide a looser comparison of values, to verify equivalence rather than more strict equality. For example, you can do this with anonymous types:

```csharp
public class Foo
{
    public int Bar { get; set; }
}

[Fact]
public void Example()
{
    var expected = new     { Bar = 42 };
    var actual   = new Foo { Bar = 42 };

    Assert.Equivalent(expected, actual);
}
```

You can see that these are two different types, but they contain a value that can be said to be equivalent. `Assert.Equal` would fail, while `Assert.Equivalent` passes.

You will need to use 2.5.1 for this to work, because of a [bug related to how IComparable was handled causing failures with Tuples](https://github.com/xunit/xunit/issues/2758) in Assert.Equivalent. (Always tuples causing me so much trouble. 😂) So taking the two original tests above and rewriting them to use `Assert.Equivalent` allows them both to pass:

```csharp
public class UnitTest
{
    [Fact]
    public void Tuple()
    {
        var expected = (1, new[] { 2 });
        var actual   = (1, new[] { 2 });

        Assert.Equivalent(expected, actual);
    }

    [Fact]
    public void TupleInCollection()
    {
        var expected = new[] { (1, new[] { 2 }) };
        var actual   = new[] { (1, new[] { 2 }) };

        Assert.Equivalent(expected, actual);
    }
}

```

```
xUnit.net Console Runner v2.5.1-pre.22+c6073bb042 (64-bit .NET Framework 4.7.2, runtime: 4.0.30319.42000)
  Discovering: empty (app domain = on [shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  empty (found 2 test cases)
  Starting:    empty (parallel test collections = off, max threads = 24)
  Finished:    empty
=== TEST EXECUTION SUMMARY ===
   empty  Total: 2, Errors: 0, Failed: 0, Skipped: 0, Time: 0.155s
```

2.5.1 has not reached a final release yet, but a release candidate quality build is available on NuGet ([`2.5.1-pre.20`](https://www.nuget.org/packages/xunit/2.5.1-pre.20)) and I have verified that it does work as expected there. (My examples show `2.5.1-pre.22` because I'm using the build from CI, and there has been a bug fix since that last shipment to NuGet to fix an issue unrelated to your problem.)

**Reply by toore** [reply]
Thanks for the thorough explanation! 🤩 I agree, sounds reasonably. 

I've upgraded to `2.5.1-pre.20` and changed the "failing" asserts to using `Equivalent` instead. Works like a charm! Thanks!
